test/integration/replicaset/replicaset_test.go

```diff
@@ -268,250 +493,454 @@ func TestAdoption(t *testing.T) {
+	rs := newRS("rs", ns.Name, 1)
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	podClient := c.Core().Pods(ns.Name)
+	pods := getPods(t, podClient, labelMap())
+	if len(pods.Items) != 1 {
+		t.Fatalf("len(pods) = %d, want 1", len(pods.Items))
+	}
+	pod := &pods.Items[0]
+	var falseVar = false
+
+	// When the only OwnerReference of the pod points to another type of API object such as statefulset
+	// with Controller=false, the RS should add a second OwnerReference (ControllerRef) pointing to itself
+	// with Controller=true
+	ownerReference := metav1.OwnerReference{UID: uuid.NewUUID(), APIVersion: "apps/v1beta1", Kind: "StatefulSet", Name: rs.Name, Controller: &falseVar}
+	testPodControllerRefPatch(t, c, pod, &ownerReference, rs, 2)
+
+	// When the only OwnerReference of the pod points to the RS, but Controller=false
+	ownerReference = metav1.OwnerReference{UID: rs.UID, APIVersion: "extensions/v1beta1", Kind: "ReplicaSet", Name: rs.Name, Controller: &falseVar}
+	testPodControllerRefPatch(t, c, pod, &ownerReference, rs, 1)
+}
 
-	// change the rs's selector to match both pods
-	patch := `{"metadata":{"labels":{"name":null}}}`
-	podClient := clientSet.Core().Pods(ns.Name)
-	pod2, err := podClient.Patch(pod2.Name, types.StrategicMergePatchType, []byte(patch))
+func TestReadyAndAvailableReplicas(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-ready-and-available-replicas", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	stopCh := runControllerAndInformers(t, rm, informers, 0)
+	defer close(stopCh)
+
+	rs := newRS("rs", ns.Name, 3)
+	rs.Spec.MinReadySeconds = 30
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	// First verify no pod is available
+	if rs.Status.AvailableReplicas != 0 {
+		t.Fatalf("Unexpected .Status.AvailableReplicas: Expected 0, saw %d", rs.Status.AvailableReplicas)
+	}
+
+	podClient := c.Core().Pods(ns.Name)
+	pods := getPods(t, podClient, labelMap())
+	if len(pods.Items) != 3 {
+		t.Fatalf("len(pods) = %d, want 3", len(pods.Items))
+	}
+
+	// Separate 3 pods into their own list
+	firstPodList := &v1.PodList{Items: pods.Items[:1]}
+	secondPodList := &v1.PodList{Items: pods.Items[1:2]}
+	thirdPodList := &v1.PodList{Items: pods.Items[2:]}
+	// First pod: Running, but not Ready
+	// by setting the Ready condition to false with LastTransitionTime to be now
+	setPodsStatus(t, c, firstPodList, v1.ConditionFalse, time.Now())
+	// Second pod: Running and Ready, but not Available
+	// by setting LastTransitionTime to 30 minutes ago
+	setPodsStatus(t, c, secondPodList, v1.ConditionTrue, time.Now().Add(-30*time.Minute))
+	// Third pod: Running, Ready, and Available
+	// by setting LastTransitionTime to now
+	setPodsStatus(t, c, thirdPodList, v1.ConditionTrue, time.Now())
+	waitPodsReady(t, c, 2, rs)
```

Comment:
Since you don't set the first Pod as Ready, waiting for the RS to show `Status.ReadyReplicas == 2` doesn't guarantee that the RS has observed the change to the first Pod.

Instead, combine this and the checks below into one `PollImmediate` that verifies all the Status values together (Replicas, ReadyReplicas, AvailableReplicas).

Reply:
Agreed that using `PollImmediate` is the correct way to verify all changes simultaneously.